File: console/src/export/download.ts
// Copyright 2026 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { type Synnax as Client } from "@synnaxlabs/client";
import { type Status } from "@synnaxlabs/pluto";
import { URL } from "@synnaxlabs/x";

import { type Cluster } from "@/cluster/slice";
import { downloadStream } from "@/runtime/download";

export interface BackupExportRequest {
  workspace_keys?: string[];
  user_keys?: string[];
  device_keys?: string[];
  task_keys?: string[];
  range_keys?: string[];
  channel_keys?: string[];
}

export interface DownloadBackupParams {
  client: Client;
  cluster: Cluster;
  request: BackupExportRequest;
  addStatus: Status.Adder;
  onDownloadStart?: () => void;
}

export const downloadBackup = async ({
  client,
  cluster,
  request,
  addStatus,
  onDownloadStart,
}: DownloadBackupParams): Promise<void> => {
  if (!client.auth.authenticated)
    throw new Error("Client is not authenticated");

  const url = new URL({
    host: cluster.host,
    port: Number(cluster.port),
    protocol: cluster.secure ? "https" : "http",
    pathPrefix: "api/v1",
  });

  const response = await fetch(url.child("export").toString(), {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${client.auth.token}`,
    },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    const body = await response.text();
    throw new Error(`Export failed (${response.status}): ${body}`);
  }

  const stream = response.body;
  if (stream == null) throw new Error("Export response has no body");

  await downloadStream({
    stream,
    name: "backup",
    extension: "sy",
    addStatus,
    onDownloadStart,
  });
};

File: freighter/py/freighter/download.py
#  Copyright 2026 Synnax Labs, Inc.
#
#  Use of this software is governed by the Business Source License included in the file
#  licenses/BSL.txt.
#
#  As of the Change Date specified in that file, in accordance with the Business Source
#  License, use of this software will be governed by the Apache License, Version 2.0,
#  included in the file licenses/APL.txt.

from __future__ import annotations

from typing import Protocol

from freighter.transport import RQ, FilePath, Transport


class DownloadClient(Transport, Protocol):
    """
    Protocol for sending a typed request RQ and streaming the response directly into a
    destination file. Used when the response can be too large to buffer in memory. The
    transport infers any wire-format metadata from the destination (e.g., from its
    extension).
    """

    def download(
        self,
        target: str,
        req: RQ,
        dest: FilePath,
    ) -> None:
        """
        Sends req to target and streams the response body into dest.

        :param target: the target address of the server.
        :param req: the typed request payload.
        :param dest: file path to stream the response body into.
        :raises Unreachable: when the target cannot be reached.
        :raises Exception: any error returned by the server.
        """
        ...
